### PR TITLE
Cm forest discovery

### PR DESCRIPTION
**Pull request by NEllis280**

### CODE REVIEW

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+    Import-ConfigMgrPowerShellModule -SiteCode $SiteCode
+    Set-Location -Path "$($SiteCode):\"
+
+    try
+    {
+        $state = Get-TargetResource -SiteCode $SiteCode -Enabled $Enabled
+        if ($Enabled -eq $true)
+        {
+            $exludelist = @('Verbose','PollingSchedule')
```
Comment (rchristman89) [inline source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1:116]: I would change this to be an includelist, there are other commonparameters that could be used that will cause an error.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+                    RecurCount    = $state.PollingSchedule.RecurCount
+                }
+
+                $stateSchedule = New-CMSchedule @currentSchedule
+
+                $array = @('DayDuration','DaySpan','HourDuration','HourSpan','IsGMT','MinuteDuration','MinuteSpan')
+
+                foreach ($item in $array)
+                {
+                    if (($desiredPollingSchedule).$($item) -ne ($stateSchedule).$($item))
```
Comment (rchristman89) [inline source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1:151]: This: if (($desiredPollingSchedule).$($item) -ne ($stateSchedule).$($item))
Could be: if ($desiredPollingSchedule.$($item) -ne $stateSchedule.$($item))

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+                    }
+                }
+            }
+
+            if ($buildingParams)
+            {
+                Set-CMDiscoveryMethod -ActiveDirectoryForestDiscovery -SiteCode $SiteCode @buildingParams
+            }
+        }
+        else
```
Comment (rchristman89) [inline source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1:172]: This could be changed to an elseif ($Enabled -eq $true -and $state.Enabled -eq $true)

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    }
+    catch
+    {
+        throw $_
+    }
+    finally
+    {
+        Set-Location -Path "$env:temp"
+    }
+}
```
Comment (rchristman89) [inline source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1:188]: Need an extra line below.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    )
+
+    Import-ConfigMgrPowerShellModule -SiteCode $SiteCode
+    Set-Location -Path "$($SiteCode):\"
+    $state = Get-TargetResource -SiteCode $SiteCode -Enabled $Enabled
+    $result = $true
+
+    if ($Enabled -eq $true)
+    {
+        $exludelist = @('Verbose','PollingSchedule')
```
Comment (rchristman89) [inline source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1:244]: Same as above I would change this to an include list.

File: tests/Unit/DSC_CMForestDiscovery.tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,259 @@
+[System.Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSAvoidUsingConvertToSecureStringWithPlainText', '')]
```
Comment (rchristman89) [inline tests/Unit/DSC_CMForestDiscovery.tests.ps1:1]: This line isn't needed.

File: tests/Unit/DSC_CMForestDiscovery.tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        $getInputDisableSubnet = @{
+            SiteCode                     = 'Lab'
+            Enabled                      = $true
+            EnableSubnetBoundaryCreation = $false
+        }
+        $getReturnDisabled = @{
+            SiteCode = 'Lab'
+            Enabled  = $false
+        }
+        Describe "$moduleResourceName\Get-TargetResource" {
```
Comment (rchristman89) [inline tests/Unit/DSC_CMForestDiscovery.tests.ps1:144]: Take a look at other Unit tests and fixing spacing based on those.

File: tests/Unit/DSC_CMForestDiscovery.tests.ps1 [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+            Mock -CommandName Import-ConfigMgrPowerShellModule
+            Mock -CommandName Set-Location
+            Context 'When retrieving Collection settings' {
+                It 'Should return desired result for forest discovery.' {
+                    mock -CommandName Get-CMDiscoveryMethod -MockWith { $standardGetDiscoveryOutput }
```
Comment (rchristman89) [inline tests/Unit/DSC_CMForestDiscovery.tests.ps1:149]: Casing: Mock

File: tests/Unit/DSC_CMForestDiscovery.tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+                    Set-TargetResource @getReturnDisabled
+                    Assert-MockCalled Import-ConfigMgrPowerShellModule -Exactly -Times 1 -Scope It
+                    Assert-MockCalled Set-Location -Exactly -Times 2 -Scope It
+                    Assert-MockCalled Get-TargetResource -Exactly -Times 1 -Scope It
+                    Assert-MockCalled New-CMSchedule -Exactly -Times 0 -Scope It
+                    Assert-MockCalled Set-CMDiscoveryMethod -Exactly -Times 1 -Scope It
+                }
+            }
+            Context 'When running Set-TargetResource should throw' {
+                It 'Should call expected commands and throw if query membership throws' {
```
Comment (rchristman89) [inline tests/Unit/DSC_CMForestDiscovery.tests.ps1:200]: Query Membership?

File: tests/Unit/DSC_CMForestDiscovery.tests.ps1 (hunk 8 above)
Comment (NEllis280) [reply tests/Unit/DSC_CMForestDiscovery.tests.ps1]: Changed

File: tests/Unit/DSC_CMForestDiscovery.tests.ps1 (hunk 6 above)
Comment (NEllis280) [reply tests/Unit/DSC_CMForestDiscovery.tests.ps1]: Removed

File: source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1 [hunk 10]
```diff
@@ -548,6 +548,45 @@ function Convert-CidrToIP
...
+    [OutputType([Microsoft.Management.Infrastructure.CimInstance])]
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [String]
+        $ScheduleString,
+        [Parameter(Mandatory = $true)]
+        [String]
+        $CimClassName
+    )
```
[the remark this answers is not in the document]
Comment (NEllis280) [reply source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1:564]: Added whitespace.

File: source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1 [hunk 11]
```diff
@@ -548,6 +548,45 @@ function Convert-CidrToIP
...
 
+function ConvertTo-CimCMScheduleString
+{
+    [CmdletBinding()]
+    [OutputType([Microsoft.Management.Infrastructure.CimInstance])]
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [String]
+        $ScheduleString,
```
[the remark this answers is not in the document]
Comment (NEllis280) [reply source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1:559]: Added.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.schema.mof [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    [Write, Description("Indicates whether Configuration Manager creates IP address range boundaries from AD DS discovery information.")] Boolean EnableSubnetBoundaryCreation;
+};
+
+[ClassVersion("1.0.0")]
+class DSC_CMForestDiscoveryPollingSchedule
+{
+     [Required, Description("Specifies the recur interval of days, hours, or minutes."),ValueMap{"Days","Hours","Minutes"},Values{"Days","Hours","Minutes"}] String RecurInterval;
+     [Required, Description("Specifies how often the recur interval is run.")] UInt32 RecurCount;
+};
+
```
[the remark this answers is not in the document]
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.schema.mof]: Removed.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.schema.mof [hunk 13]
```diff
@@ -0,0 +1,17 @@
+[ClassVersion("1.0.0"), FriendlyName("CMForestDiscovery")]
+class DSC_CMForestDiscovery: OMI_BaseResource
+{
+    [Key, Description("Specifies the SiteCode for the Configuration Manager site.")] String SiteCode;
+    [Required, Description("Specifies the enablement of the forest discovery method. If settings is set to $false no other value provided will be evaluated for compliance.")] Boolean Enabled;
+    [Write, EmbeddedInstance("DSC_CMForestDiscoveryPollingSchedule"), Description("Hashtable containing polling schedule for Configuration Manager (RecurInterval, RecurCount)")] String PollingSchedule;
```
[the remark this answers is not in the document]
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.schema.mof]: Removed

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 (hunk 4 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1:188]: Added.

File: source/ConfigMgrCBDsc.psd1 [hunk 14]
```diff
@@ -47,6 +47,7 @@
         'SccmIniFile'
         'CMCollections'
         'CMBoundaries'
+        'DSC_CMForestDiscovery'
```
[the remark this answers is not in the document]
Comment (NEllis280) [reply source/ConfigMgrCBDsc.psd1]: Removed

File: README.md [hunk 15]
```diff
@@ -216,3 +218,24 @@ Please check out common DSC Community [contributing guidelines](https://dsccommu
...
 - [CMBoundaries_Present](Source\Examples\Resources\CMBoundaries\CMBoundaries_Present.ps1)
+
+### CMForestDiscovery
+
+- **[String] SiteCode** _(Key)_: Specifies the Site Code for the Configuration
+  Manager site.
+- **[Boolean] Enabled** _(Required)_: Specifies the enablement of the forest
+  discovery method. If settings is set to $false no other value provided will be
+  evaluated for compliance.
+- **[EmbeddedInstance] PollingSchedule** _(Write)_: Hashtable containing polling
```
[the remark this answers is not in the document]
Comment (NEllis280) [reply README.md]: Resolved

File: tests/Unit/DSC_CMForestDiscovery.tests.ps1 (hunk 9 above)
Comment (NEllis280) [reply tests/Unit/DSC_CMForestDiscovery.tests.ps1]: Good catch, this was left over from an example I used to set up throws. Fixed here.

File: tests/Unit/DSC_CMForestDiscovery.tests.ps1 (hunk 7 above)
Comment (NEllis280) [reply tests/Unit/DSC_CMForestDiscovery.tests.ps1:157]: Fixed, based on the spacing in the other unit tests.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 (hunk 2 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1]: Changed in both test and set.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 (hunk 3 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1]: updated, except we want it to be elseif ($Enabled -eq $false -and $state.Enabled -eq $true)

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 (hunk 1 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1]: Swapped out for an includeList.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 (hunk 5 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1]: Swapped out for an includeList.

File: source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1 [hunk 16]
```diff
@@ -548,6 +548,50 @@ function Convert-CidrToIP
...
+
+        [Parameter(Mandatory = $true)]
+        [String]
+        $CimClassName
+    )
+
+    $schedule = Convert-CMSchedule -ScheduleString $ScheduleString
+
+    if (-not [string]::IsNullOrEmpty($schedule.DaySpan))
```
Comment (rchristman89) [inline source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1:568]: Should this be something different then , $schedule.DaySpan? I know that it will be there if it is a valid string but it read that it is specific to days.

File: source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1 [hunk 17]
```diff
@@ -548,6 +548,50 @@ function Convert-CidrToIP
...
+    {
+        if ($schedule.DaySpan -gt 0)
```
Comment (rchristman89) [inline source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1:570]: Can $schedule.DaySpan, HourSpan, or MinuteSPan coexist?

File: source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1 (hunk 16 above)
Comment (jeffotterpohl) [reply source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1:568]: This is just a check to validate that the schedule isn't set to None.  If it is set to none DaySpan, HoureSpan, and MinutesSpan doesn't exist in the schedule

File: source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1 (hunk 17 above)
Comment (jeffotterpohl) [reply source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1:570]: you can't mix and match, you can't have hours and minutes, or Days, hours, minutes. it is a chose one type option

File: tests/Unit/ConfigMgrCBDsc.ResourceHelper.tests.ps1 [hunk 18]
```diff
@@ -765,4 +765,76 @@ InModuleScope $script:subModuleName {
...
+            HourSpan       = 0
+            MinuteDuration = 0
+            MinuteSpan     = 50
+        }
+        $cimInputParam = @{
+            CimClassName   = 'DSC_TestCimInstance'
+            ScheduleString = '0001200000100038'
+        }
+        Context 'When return is as expected' {
+            It 'Should return desired result for hearbeat discovery hours' {
```
Comment (rchristman89) [inline tests/Unit/ConfigMgrCBDsc.ResourceHelper.tests.ps1:820]: The It block says the same thing for all three. Is this directly related to discovery hours? 

Also, I know it works by using the same input for since the function is being mocked but it would be nice to have the different inputs for testing that reflect the outputs being mocked.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.schema.mof [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+    [Required, Description("Specifies the enablement of the forest discovery method. If settings is set to $false no other value provided will be evaluated for compliance.")] Boolean Enabled;
+    [Write, EmbeddedInstance("DSC_CMForestDiscoveryPollingSchedule"), Description("Contains the polling schedule for Configuration Manager (RecurInterval, RecurCount)")] String PollingSchedule;
+    [Write, Description("Indicates whether Configuration Manager creates Active Directory boundaries from AD DS discovery information.")] Boolean EnableActiveDirectorySiteBoundaryCreation;
+    [Write, Description("Indicates whether Configuration Manager creates IP address range boundaries from AD DS discovery information.")] Boolean EnableSubnetBoundaryCreation;
+};
+
+[ClassVersion("1.0.0")]
+class DSC_CMForestDiscoveryPollingSchedule
+{
+     [Required, Description("Specifies the recur interval of days, hours, or minutes."),ValueMap{"Days","Hours","Minutes"},Values{"Days","Hours","Minutes"}] String RecurInterval;
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.schema.mof:14]: Minutes are not allowed in Forest Discovery

File: source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1 [hunk 20]
```diff
@@ -548,6 +548,50 @@ function Convert-CidrToIP
     }
 }
 
+function ConvertTo-CimCMScheduleString
```
Comment (jeffotterpohl) [inline source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1:551]: should have a synopsis

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+    .PARAMETER PollingSchedule
+        Specifies a schedule and determines how often Configuration Manager attempts to discover Active Directory forest.
+
+    .PARAMETER EnableActiveDirectorySiteBoundaryCreation
+        Indicates whether Configuration Manager creates Active Directory boundaries from AD DS discovery information.
+
+    .PARAMETER EnableSubnetBoundaryCreation
+        Indicates whether Configuration Manager creates IP address range boundaries from AD DS discovery information.
+#>
+
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1:207]: extra line

File: CHANGELOG.md [hunk 22]
```diff
@@ -14,6 +14,7 @@ For older change log history see the [historic changelog](HISTORIC_CHANGELOG.md)
 - Added Set-ConfigMgrCert to the ResourceHelper
 - Added CMBoundaries resource
 - Added Convert-CidrToIP to the ResourceHelper
+- Added CMForestDiscovery resource
```
Comment (jeffotterpohl) [inline CHANGELOG.md:17]: also added ConvertTo-CimCMScheduleString to the ResourceHelper

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 (hunk 21 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1:207]: Removed Extra line

File: CHANGELOG.md (hunk 22 above)
Comment (NEllis280) [reply CHANGELOG.md:17]: Added to Changelog

File: source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1 (hunk 20 above)
Comment (NEllis280) [reply source/Modules/ConfigMgrCBDsc.ResourceHelper/ConfigMgrCBDsc.ResourceHelper.psm1:561]: Synopsis added

File: tests/Unit/ConfigMgrCBDsc.ResourceHelper.tests.ps1 (hunk 18 above)
Comment (NEllis280) [reply tests/Unit/ConfigMgrCBDsc.ResourceHelper.tests.ps1]: Changed the descriptions. Will address the rest soon.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.schema.mof (hunk 19 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.schema.mof]: Removed from schema and array (in the psm1)

File: tests/Unit/ConfigMgrCBDsc.ResourceHelper.tests.ps1 (hunk 18 above)
Comment (NEllis280) [reply tests/Unit/ConfigMgrCBDsc.ResourceHelper.tests.ps1]: I gave it different inputs, but since you said it may be redundant, I didn't do any assert mocks

Comment (rchristman89) [reply tests/Unit/ConfigMgrCBDsc.ResourceHelper.tests.ps1]: It may be good practice to do the Assert since we are saying that mock should be ran in all the scenarios. I would verify with Jeff his thoughts.

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+    )
+
+    Write-Verbose -Message $script:localizedData.RetrieveSettingValue
+    Import-ConfigMgrPowerShellModule -SiteCode $SiteCode
+    Set-Location -Path "$($SiteCode):\"
+
+    $forestDiscovery = (Get-CMDiscoveryMethod -Name ActiveDirectoryForestDiscovery -SiteCode $SiteCode).Props
+    $enabledStatus   = (($forestDiscovery | Where-Object -FilterScript {$_.PropertyName -eq 'Settings'}).Value1 -eq 'Active')
+    $forestSchedule  = ($forestDiscovery | Where-Object -FilterScript {$_.PropertyName -eq 'Startup Schedule'}).Value1
+    $subnetboundary  = ($forestDiscovery | Where-Object -FilterScript {$_.PropertyName -eq 'Enable Subnet Boundary Creation'}).Value
```
[the remark this answers is not in the document]
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1]: Fixed capitalization

File: source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1 [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+    $siteboundary    = ($forestDiscovery | Where-Object -FilterScript {$_.PropertyName -eq 'Enable AD Site Boundary Creation'}).Value
```
[the remark this answers is not in the document]
Comment (NEllis280) [reply source/DSCResources/DSC_CMForestDiscovery/DSC_CMForestDiscovery.psm1]: Fixed capitalization

File: tests/Unit/ConfigMgrCBDsc.ResourceHelper.tests.ps1 (hunk 18 above)
Comment (NEllis280) [reply tests/Unit/ConfigMgrCBDsc.ResourceHelper.tests.ps1]: I spoke to Jeff and got this switched over.
